File: 3주차/BOJ2667. 단지번호붙이기/최원재.py
dx = [-1, 1 ,0, 0]
dy = [0, 0, -1, 1]

def bfs(i, j):
    q = []
    visited[i][j] = True
    q.append([i,j])
    cnt = 0
    while(q):
        cnt+=1
        cur = q.pop()
        for i in range(4):
            nr = cur[0] + dx[i]
            nc = cur[1] + dy[i]
            if arr[nr][nc] is '1' and not visited[nr][nc]:
                visited[nr][nc] = True
                q.append([nr,nc])
    result.append(cnt)

N = int(input())
arr = []
visited = [[False for j in range(N+2)] for i in range(N+2)]
arr.append(['0' for i in range(N+2)])
for _ in range(N):
    arr.append('0'+input()+'0')
arr.append(['0' for i in range(N+2)])

result = []

for i in range(1,N+1):
    for j in range(1,N+1):
        if arr[i][j] is '1' and not visited[i][j]:
            bfs(i, j)
result.sort()
length = len(result)
print(length)
for i in range(length):
    print(result[i])

File: 3주차/BOJ2667. 단지번호붙이기/최원재.cpp
#include <iostream>
#include <vector>
#include <string>
#include <queue>

using namespace std;

vector<string> v;
vector<vector<bool>> visited;
int n;
int dx[4] = { -1, 1, 0, 0 };
int dy[4] = { 0, 0, -1, 1 };

int bfs(int x, int y) {
	queue<pair<int,int>> q;
	visited[x][y] = true;
	q.push(make_pair(x, y));
	pair<int, int> cur;
	int nx, ny;
	int cnt = 0;
	while (!q.empty()){
		cnt++;
		cur = q.front();
		q.pop();
		for (int i = 0; i < 4; i++) {
			nx = cur.first + dx[i];
			ny = cur.second + dy[i];
			if (nx < 0 || nx >= n || ny < 0 || ny >= n) continue;
			if (v[nx][ny] == '1' && !visited[nx][ny]) {
				q.push(make_pair(nx, ny));
				visited[nx][ny] = true;
			}
		}
	}
	return cnt;
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cin >> n;
	visited.resize(n+1, vector<bool>(n+1));
	string s;
	for (int i = 0; i <n; ++i) {
		cin >> s;
		v.push_back(s);
	}

	priority_queue<int, vector<int>> pq;
	
	for (int i = 0; i < n; ++i) {
		for (int j = 0; j < n; ++j) {
			if (v[i][j] == '1' && !visited[i][j]) pq.push(0-bfs(i, j));
		}
	}
	int len = pq.size();
	cout << len << '\n';
	for (int i = 0; i < len; i++) {
		cout << 0 - pq.top() << '\n';
		pq.pop();
	}
}
